File: benchmarks/v3/scenarios/11-python-fastapi-repository/with-mcp/app/application/user_service.py
from typing import List
from ..domain.user import User, UserCreate, UserUpdate
from ..domain.repository import UserRepository
from ..domain.exceptions import UserNotFoundException, UserAlreadyExistsException


class UserService:
    def __init__(self, repository: UserRepository):
        self.repository = repository

    async def create_user(self, data: UserCreate) -> User:
        existing = await self.repository.get_by_email(data.email)
        if existing:
            raise UserAlreadyExistsException(data.email)
        return await self.repository.create(data)

    async def get_user(self, user_id: int) -> User:
        user = await self.repository.get_by_id(user_id)
        if not user:
            raise UserNotFoundException(user_id)
        return user

    async def get_all_users(self) -> List[User]:
        return await self.repository.get_all()

    async def update_user(self, user_id: int, data: UserUpdate) -> User:
        if data.email:
            existing = await self.repository.get_by_email(data.email)
            if existing and existing.id != user_id:
                raise UserAlreadyExistsException(data.email)

        user = await self.repository.update(user_id, data)
        if not user:
            raise UserNotFoundException(user_id)
        return user

    async def delete_user(self, user_id: int) -> None:
        deleted = await self.repository.delete(user_id)
        if not deleted:
            raise UserNotFoundException(user_id)

File: benchmarks/v3/scenarios/06-ts-express/with-mcp/src/application/userService.ts
import { v4 as uuidv4 } from 'uuid';
import { User, CreateUserDto, UpdateUserDto } from '../domain/user.js';
import { NotFoundError, ConflictError } from '../domain/errors.js';
import { UserRepository } from '../infrastructure/repository/userRepository.js';

export interface UserService {
  create(dto: CreateUserDto): Promise<User>;
  getById(id: string): Promise<User>;
  getAll(): Promise<User[]>;
  update(id: string, dto: UpdateUserDto): Promise<User>;
  delete(id: string): Promise<void>;
  findByEmail(email: string): Promise<User | undefined>;
}

export class UserServiceImpl implements UserService {
  constructor(private readonly repository: UserRepository) {}

  async create(dto: CreateUserDto): Promise<User> {
    const existing = await this.repository.findByEmail(dto.email);
    if (existing) {
      throw new ConflictError('Email already registered');
    }

    const user: User = {
      id: uuidv4(),
      email: dto.email,
      name: dto.name,
      role: dto.role,
      passwordHash: this.hashPassword(dto.password),
      createdAt: new Date(),
    };

    return this.repository.save(user);
  }

  async getById(id: string): Promise<User> {
    const user = await this.repository.findById(id);
    if (!user) {
      throw new NotFoundError('User');
    }
    return user;
  }

  async getAll(): Promise<User[]> {
    return this.repository.findAll();
  }

  async update(id: string, dto: UpdateUserDto): Promise<User> {
    const user = await this.getById(id);

    if (dto.email && dto.email !== user.email) {
      const existing = await this.repository.findByEmail(dto.email);
      if (existing) {
        throw new ConflictError('Email already registered');
      }
    }

    const updated: User = {
      ...user,
      email: dto.email ?? user.email,
      name: dto.name ?? user.name,
      role: dto.role ?? user.role,
    };

    return this.repository.save(updated);
  }

  async delete(id: string): Promise<void> {
    await this.getById(id);
    await this.repository.delete(id);
  }

  async findByEmail(email: string): Promise<User | undefined> {
    return this.repository.findByEmail(email);
  }

  private hashPassword(password: string): string {
    return Buffer.from(password).toString('base64');
  }
}
